text/0232-maturity-of-otel.md

```diff
@@ -0,0 +1,69 @@
+### Maturity levels
+
+Deliverables of a SIG MUST have a declared maturity level, established by SIG maintainers (SIGs), likely with the input of the code owners. While the main deliverable can have a specific maturity level, individual components might have a different one. Examples:
+* the Collector core distribution might declare itself stable and include a receiver that is not stable. In that case, the receiver has to be clearly marked as such
+* the Java Agent might be declared stable, while individual instrumentation packages are not
+
+It is recommended that components should not be marked as stable if their user-visible interfaces are not stable. For instance, if the Collector's component "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" should be at most "beta". Maintainers are free to deviate from this recommendation if they believe users are not going to be affected by future changes.
+
+#### Development
+
+Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.
+
+#### Alpha
+
+This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The component's interface and configuration options might often change without backward compatibility guarantees. Components at this stage might be dropped at any time without notice.
+
+#### Beta
+
+Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.
+
+#### Feature freeze
+
+The component is feature-complete and ready for broader usage. The component is ready to be declared stable, it might just need to be tested in more production environments before that can happen. Bugs and performance problems are expected to be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.
```

Comment:
Can we get a clear definition of a breaking change as part of this document or pointing to a definition elsewhere?

Reply:
I had more concrete definitions of what's a breaking change, but it proved too difficult to get a definition that would work for everything: signals, contrib SDKs, agents, collector, ... We delegate now this to each SIG to define.

Reply:
There has be a lowest common denominator. Microsoft defines a breaking change as:

> A breaking change is a change that may require you to make changes to your application in order to avoid disruption to your integration.

Reply:
That definition works for me. Can we adopt it formally and have each SIG define what their integration surface is (guessing API, config, inputs/outputs)?

Reply:
FWIW I think this can be refined further.
> A breaking change is a change that may require you make changes to your application in order to avoid disruption to your integration, *outside normal change management through communication, deprecation and versioning*.

For example:
Breaking change
> This config option is removed

Non-breaking change
> This config option is deprecated and will be removed in 3 months or 6 releases, whichever comes first. The option is replaced by this other functionality. Please read those migration steps.

Reply:
Here is what I think we can adopt in this OTEP:

```
"A breaking change is a change that may require you to make changes to your application in order to avoid disruption to your integration."
Microsoft, [LinkedIn breaking change policy](https://learn.microsoft.com/en-us/linkedin/shared/breaking-change-policy)

To understand if a change is to be considered a breaking change, please follow this ruleset:
1. Does the change require users or developers of the component to make themselves changes to avoid disruption of the integration with your component?
   If YES, continue. If NO, the change is considered a non-breaking change.
2. Was the change communicated in advance as per the guidelines of the project, either through release notes, deprecation notice or documentation?
3. Was a mitigation strategy, such as migration steps or workaround offered?
4. Is the change associated with a new version which stipulates the breaking change clearly (using semantic versioning, release notes)?
   If you have replied YES to 2, 3 and 4, the change is considered a non-breaking change. Otherwise, it is considered a breaking change.
```

Reply:
While that's reasonable for a library or for the collector, that's not an acceptable approach to the protocol: clients are expected to exist for years to come without the expectation that they need to be changed at all. We shouldn't expect new versions of servers to require clients to adopt new versions of our protocol formats in order to communicate with the server, no matter how long in advance we communicated the change.

Reply:
That said, I do like the original definition: "A breaking change is a change that may require you to make changes to your application in order to avoid disruption to your integration."

Perhaps:

"A breaking change is a change that may require consumers of our components to adapt themselves in order to avoid disruption to their usage of our components."

Some examples of breaking changes:
* instrumented applications that need to be changed because the API has been changed
* the SDK starts collecting substantially more telemetry because of a new behavior
* a change to OTLP requires clients to bump their dependency versions in order to communicate with servers on newer versions
* a new semantic convention is added as required

Again, breaking changes are acceptable, just not for stable components.

Reply:
The definition of what is a breaking change is important and it may warrant a separate OTEP. I think we don't necessarily need to have a precise definition of it in order to move forward with this OTEP.

Reply:
@atoulme, do you think the following is sufficient for now?

> A breaking change is a change that may require consumers of our components to adapt themselves in order to avoid disruption to their usage of our components.

Reply:
OK let’s roll

Reply:
I added this before we start defining the levels. Can you check again?

Reply:
lgtm